protobuf/protoc-gen-cruxclient/cruxclient_generator.cc

```diff
@@ -0,0 +1,426 @@
+    vars["request"] = ClassName(method->input_type(), true);
+    vars["response"] = ClassName(method->output_type(), true);
+
+    if (method->client_streaming()) {
+      // [RC]: Client Steaming not supported yet
+      continue;
+    }
+
+    if (method->server_streaming()) {
+      vars["response"] = "std::vector<" + vars["response"] + ">";
+    }
+    if (isVirtual) {
+      printer->Print("virtual ");
+    }
+    printer->Print(vars,
+                   "$response$ $method_name$(const $request$& "
+                   "request) const");
+    if (isVirtual) {
+      printer->Print(" = 0");
+    }
+    printer->Print(";\n");
+  }
+}
+
+void PrintHeaderInterfaces(Printer *printer, const FileDescriptor *file) {
+  std::map<string, string> vars;
+  for (int service_index = 0; service_index < file->service_count();
+       ++service_index) {
+    const ServiceDescriptor *service = file->service(service_index);
+    vars["service_name"] = service->name();
+
+    printer->Print(vars, "class $service_name$ClientInterface {\n");
+    printer->Print(" public:\n");
+    printer->Indent();
+    printer->Print(vars, "virtual ~$service_name$ClientInterface() {}\n");
+    PrintHeaderMethods(printer, service, true);
+    printer->Outdent();
+    printer->Print("};\n\n");
+  }
+}
+
+void PrintHeaderClients(Printer *printer, const FileDescriptor *file) {
+  std::map<string, string> vars;
+
+  for (int service_index = 0; service_index < file->service_count();
+       ++service_index) {
+    const ServiceDescriptor *service = file->service(service_index);
+    vars["service_name"] = service->name();
+
+    printer->Print(
+        vars,
+        "class $service_name$Client: public $service_name$ClientInterface {\n");
+    printer->Print(" public:\n");
+    printer->Indent();
+    printer->Print(vars,
+                   "explicit $service_name$Client(const "
+                   "std::shared_ptr<$service_name$::StubInterface>& "
+                   "stub);\n");
+    PrintHeaderMethods(printer, service, false);
+    printer->Outdent();
+    printer->Print("\n");
+
+    printer->Print(" private:\n");
+    printer->Indent();
+    printer->Print(vars,
+                   "std::shared_ptr<$service_name$::StubInterface> mStub;\n\n");
+    printer->Outdent();
+    printer->Print("};\n\n");
+  }
+}
+
+void PrintHeaderEpilogue(Printer *printer, const FileDescriptor *file) {
+  std::map<string, string> vars;
+  vars["filename"] = file->name();
+  vars["filename_identifier"] = FilenameIdentifier(file->name());
+
+  PrintNamespace(printer, file, true);
+}
+
+void PrintSourcePrologue(Printer *printer, const FileDescriptor *file) {
+  PrintPrologue(printer, file);
+  printer->Print("\n");
+}
+
+void PrintSourceIncludes(Printer *printer, const FileDescriptor *file) {
+  std::map<string, string> vars;
+  vars["filename_base"] = StripProto(file->name());
+
+  printer->Print(vars, "#include \"$filename_base$.crux.client.pb.h\"\n");
+  // TODO: [RC]: Change `service/` to `core/`
```

Comment:
Do we still need to retain this TODO?

Reply:
Once we migrate all the service code to this, we can move `service_utils.hpp` into the core folder and reference it there; this way we can remove the whole `service` folder.